sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala

```diff
@@ -299,6 +301,12 @@ object ShuffleExchangeExec {
           ascending = true,
           samplePointsPerPartitionHint = SQLConf.get.rangeExchangeSampleSizePerPartition)
       case SinglePartition => new ConstantPartitioner
+      case KeyGroupedPartitioning(expressions, n, partitionValues) =>
```

Comment:
It'd be great if we can also add a few tests in `EnsureRequirementsSuite` to check that Spark picks the right side to shuffle when one side of a join is non-key grouped partitioning and the other side is.